File: src/binstarjs03.AerialOBJ.Core/NbtNew/NbtReader.cs
using System;
using System.IO.Compression;
using System.IO;

namespace binstarjs03.AerialOBJ.Core.NbtNew;

public static class NbtReader
{
    private enum ByteHeader
    {
        Gzip = 0x1F,
        Zlib = 0x78,
        TagCompound = 0x0A,
    }

    public static INbt ReadDisk(string path)
    {
        MemoryStream ms = new(File.ReadAllBytes(path));
        return ReadStream(ms);
    }

    /// <exception cref="NbtDeserializationError"></exception>
    public static INbt ReadStream(Stream inputStream)
    {
        Stream stream = DecompressNbtStream(inputStream);
        NbtBinaryReader reader = new(stream);
        NbtType nbtType = ReadNbtType(reader);
        try
        {
            return ReadNbtSwitch(reader, nbtType, insideList: false);
        }
        catch (Exception e)
        {
            throw new NbtDeserializationError(
                $"Exception occured while parsing serialized nbt data.\n" +
                $"Exception details: {e}\n" +
                reader.GetNbtStackParseErrorAsString(), e);
        }
    }

    private static MemoryStream DecompressNbtStream(Stream inputStream)
    {
        MemoryStream decompressedStream = new();
        int byteHeader = inputStream.ReadByte();
        inputStream.Seek(-1, SeekOrigin.Current);
        switch (byteHeader)
        {
            case (int)ByteHeader.Gzip:
                using (GZipStream gZipStream = new(inputStream, CompressionMode.Decompress))
                    gZipStream.CopyTo(decompressedStream);
                break;
            case (int)ByteHeader.Zlib:
                using (ZLibStream zLibStream = new(inputStream, CompressionMode.Decompress))
                    zLibStream.CopyTo(decompressedStream);
                break;
            case (int)ByteHeader.TagCompound:
                inputStream.CopyTo(decompressedStream);
                break;
            default:
                throw new NbtUnknownCompressionMethodException(
                $"Undefined compression byte header of {byteHeader}");
        }
        decompressedStream.Position = 0;
        return decompressedStream;
    }

    private static INbt ReadNbtSwitch(NbtBinaryReader reader, NbtType nbtType, bool insideList)
    {
        string name = insideList ? "" : reader.ReadStringLengthPrefixed();
        reader.Stack.Push((nbtType, name));
        INbt nbt = nbtType switch
        {
            NbtType.NbtEnd => throw new NbtIllegalOperationException($"Cannot instantiate {NbtType.NbtEnd}"),
            NbtType.NbtByte => ReadNbtByte(reader, name),
            NbtType.NbtShort => ReadNbtShort(reader, name),
            NbtType.NbtInt => ReadNbtInt(reader, name),
            NbtType.NbtLong => ReadNbtLong(reader, name),
            NbtType.NbtFloat => ReadNbtFloat(reader, name),
            NbtType.NbtDouble => ReadNbtDouble(reader, name),
            NbtType.NbtString => ReadNbtString(reader, name),
            NbtType.NbtByteArray => ReadNbtByteArray(reader, name),
            NbtType.NbtIntArray => ReadNbtIntArray(reader, name),
            NbtType.NbtLongArray => ReadNbtLongArray(reader, name),
            NbtType.NbtCompound => ReadNbtCompound(reader, name),
            NbtType.NbtList => ReadNbtListSwitch(reader, name),
            _ => throw new NbtIllegalTypeException()
        };
        reader.Stack.Pop();
        return nbt;
    }

    private static NbtType ReadNbtType(NbtBinaryReader reader)
    {
        int type = reader.ReadByte();
        if (Enum.IsDefined(typeof(NbtType), type))
            return (NbtType)type;
        else
            throw new NbtIllegalTypeException(
                $"Illegal nbt type '{type}' at stream position {reader.Position}");
    }

    private static NbtByte ReadNbtByte(NbtBinaryReader reader, string name)
    {
        sbyte value = reader.ReadSByte();
        return new NbtByte(name, value);
    }

    private static NbtShort ReadNbtShort(NbtBinaryReader reader, string name)
    {
        short value = reader.ReadShortBE();
        return new NbtShort(name, value);
    }

    private static NbtInt ReadNbtInt(NbtBinaryReader reader, string name)
    {
        int value = reader.ReadIntBE();
        return new NbtInt(name, value);
    }

    private static NbtLong ReadNbtLong(NbtBinaryReader reader, string name)
    {
        long value = reader.ReadLongBE();
        return new NbtLong(name, value);
    }

    private static NbtFloat ReadNbtFloat(NbtBinaryReader reader, string name)
    {
        float value = reader.ReadFloatBE();
        return new NbtFloat(name, value);
    }

    private static NbtDouble ReadNbtDouble(NbtBinaryReader reader, string name)
    {
        double value = reader.ReadDoubleBE();
        return new NbtDouble(name, value);
    }

    private static NbtString ReadNbtString(NbtBinaryReader reader, string name)
    {
        string value = reader.ReadStringLengthPrefixed();
        return new NbtString(name, value);
    }

    private static NbtByteArray ReadNbtByteArray(NbtBinaryReader reader, string name)
    {
        int arrayLength = reader.ReadIntBE();
        sbyte[] array = new sbyte[arrayLength];
        for (int i = 0; i < array.Length; i++)
            array[i] = reader.ReadSByte();
        return new NbtByteArray(name, array);
    }

    private static NbtIntArray ReadNbtIntArray(NbtBinaryReader reader, string name)
    {
        int arrayLength = reader.ReadIntBE();
        int[] array = new int[arrayLength];
        for (int i = 0; i < array.Length; i++)
            array[i] = reader.ReadIntBE();
        return new NbtIntArray(name, array);
    }

    private static NbtLongArray ReadNbtLongArray(NbtBinaryReader reader, string name)
    {
        int arrayLength = reader.ReadIntBE();
        long[] array = new long[arrayLength];
        for (int i = 0; i < array.Length; i++)
            array[i] = reader.ReadLongBE();
        return new NbtLongArray(name, array);
    }

    private static NbtCompound ReadNbtCompound(NbtBinaryReader reader, string name)
    {
        NbtCompound nbtCompound = new(name);
        while (true)
        {
            NbtType nbtType = ReadNbtType(reader);
            if (nbtType == NbtType.NbtEnd)
                break;
            INbt nbt = ReadNbtSwitch(reader, nbtType, insideList: false);
            nbtCompound.Add(nbt.Name, nbt);
        }
        return nbtCompound;
    }

    private static INbt ReadNbtListSwitch(NbtBinaryReader reader, string name)
    {
        NbtType listType = ReadNbtType(reader);
        return listType switch
        {
            NbtType.NbtEnd => ReadNbtList<NbtByte>(reader, name, listType), // NbtEnd does not exist
            NbtType.NbtByte => ReadNbtList<NbtByte>(reader, name, listType),
            NbtType.NbtShort => ReadNbtList<NbtShort>(reader, name, listType),
            NbtType.NbtInt => ReadNbtList<NbtInt>(reader, name, listType),
            NbtType.NbtLong => ReadNbtList<NbtLong>(reader, name, listType),
            NbtType.NbtFloat => ReadNbtList<NbtFloat>(reader, name, listType),
            NbtType.NbtDouble => ReadNbtList<NbtDouble>(reader, name, listType),
            NbtType.NbtString => ReadNbtList<NbtString>(reader, name, listType),
            NbtType.NbtByteArray => ReadNbtList<NbtByteArray>(reader, name, listType),
            NbtType.NbtIntArray => ReadNbtList<NbtIntArray>(reader, name, listType),
            NbtType.NbtLongArray => ReadNbtList<NbtLongArray>(reader, name, listType),
            NbtType.NbtList => ReadNestedNbtList(reader, name),
            NbtType.NbtCompound => ReadNbtList<NbtCompound>(reader, name, listType),
            _ => throw new NbtIllegalTypeException()
        };
    }

    private static INbt ReadNestedNbtList(NbtBinaryReader reader, string name)
    {
        /* Reading nested NbtList is tricky because we can't determine what the
         * type of the NbtList is (can't instantiate, undefined type argument)
         * until we actually read what items that instance of NbtList is holding.
         *
         * To solve this, we have to polymorph it to its top-level inheritance, 
         * which is the interface of INbt in this case.
         * 
         * Technically we can inline the call in ReadNbtListSwitch at switch NbtType.NbtList
         * with => ReadNbtList<INbt>, but else we cannot make a documentation comment like this :)
         */

        // return ReadNbtList<???>(reader, name, NbtType.NbtList);
        // Here the ??? of T is Polymorphed to INbt.
        // OOP can be very powerful if done properly
        return ReadNbtList<INbt>(reader, name, NbtType.NbtList);
    }

    private static NbtList<T> ReadNbtList<T>(NbtBinaryReader reader, string name, NbtType listType) where T : class, INbt
    {
        int listLength = reader.ReadIntBE();
        NbtList<T> nbtList = new(name);
        for (int i = 0; i < listLength; i++)
        {
            T nbt = (ReadNbtSwitch(reader, listType, insideList: true) as T)!;
            nbtList.Add(nbt);
        }
        return nbtList;
    }
}

File: src/binstarjs03.AerialOBJ.CoreNative/NbtFormat/NbtReader.cpp
#include "../IO/BinaryReader.h"
#include "Nbt.h"

Nbt ReadNbtSwitch(BinaryReader& reader, Endianness endian, NbtType type, bool insideList);

NbtType ReadNbtType(BinaryReader& reader) {
    int8_t type = reader.ReadInt8();
    if (type >= 0 && type <= 12)
        return (NbtType)type;
    else {
        std::string errmsg = "Unknown Nbt type ";
        errmsg.push_back((char)type);
        throw errmsg;
    }
}

NbtByte ReadNbtByte(BinaryReader& reader, std::string name) {
    NbtByte result(name);
    result.Value = reader.ReadInt8();
    return result;
}

NbtShort ReadNbtShort(BinaryReader& reader, std::string name, Endianness endian) {
    NbtShort result(name);
    result.Value = reader.ReadInt16(endian);
    return result;
}

NbtInt ReadNbtInt(BinaryReader& reader, std::string name, Endianness endian) {
    NbtInt result(name);
    result.Value = reader.ReadInt32(endian);
    return result;
}

NbtLong ReadNbtLong(BinaryReader& reader, std::string name, Endianness endian) {
    NbtLong result(name);
    result.Value = reader.ReadInt64(endian);
    return result;
}

NbtFloat ReadNbtFloat(BinaryReader& reader, std::string name, Endianness endian) {
    NbtFloat result(name);
    result.Value = reader.ReadFloatSingle(endian);
    return result;
}

NbtDouble ReadNbtDouble(BinaryReader& reader, std::string name, Endianness endian) {
    NbtDouble result(name);
    result.Value = reader.ReadFloatDouble(endian);
    return result;
}

NbtString ReadNbtString(BinaryReader& reader, std::string name, Endianness endian) {
    NbtString result(name);
    int16_t length = reader.ReadInt16(endian);
    result.Value = reader.ReadStringUTF8(length);
    return result;
}

NbtByteArray ReadNbtByteArray(BinaryReader& reader, std::string name, Endianness endian) {
    NbtByteArray result(name);
    int32_t length = reader.ReadInt32(endian);
    for (int32_t i = 0; i < length; i++)
        result.Values.push_back(reader.ReadInt8());
    return result;
}

NbtIntArray ReadNbtIntArray(BinaryReader& reader, std::string name, Endianness endian) {
    NbtIntArray result(name);
    int32_t length = reader.ReadInt32(endian);
    for (int32_t i = 0; i < length; i++)
        result.Values.push_back(reader.ReadInt32(endian));
    return result;
}

NbtLongArray ReadNbtLongArray(BinaryReader& reader, std::string name, Endianness endian) {
    NbtLongArray result(name);
    int32_t length = reader.ReadInt32(endian);
    for (int32_t i = 0; i < length; i++)
        result.Values.push_back(reader.ReadInt64(endian));
    return result;
}

NbtCompound ReadNbtCompound(BinaryReader& reader, std::string name, Endianness endian) {
    NbtCompound result(name);
    NbtType type = ReadNbtType(reader);
    while (type != NbtTypeEnd)
    {
        Nbt nbt = ReadNbtSwitch(reader, endian, type, false);
        result.Nbts.insert(std::make_pair(nbt.Name, nbt));
        type = ReadNbtType(reader);
    }
    return result;
}

NbtList ReadNbtList(BinaryReader& reader, std::string name, Endianness endian) {
    NbtList result(name);
    NbtType listType = ReadNbtType(reader);
    int32_t length = reader.ReadInt32(endian);
    for (int32_t i = 0; i < length; i++)
    {
        Nbt nbt = ReadNbtSwitch(reader, endian, listType, true);
        result.Nbts.push_back(nbt);
    }
    return result;
}

Nbt ReadNbtSwitch(BinaryReader& reader, Endianness endian, NbtType type, bool insideList) {
    std::string name;
    if (insideList)
        name = "";
    else {
        uint16_t length = reader.ReadUInt16(endian);
        name = reader.ReadStringUTF8(length);
    }

    switch (type)
    {
        case NbtTypeEnd:
            throw "Invalid state, trying to instantiate NbtEnd";
            break;
        case NbtTypeByte:
            return ReadNbtByte(reader, name);
        case NbtTypeShort:
            return ReadNbtShort(reader, name, endian);
        case NbtTypeInt:
            return ReadNbtInt(reader, name, endian);
        case NbtTypeLong:
            return ReadNbtLong(reader, name, endian);
        case NbtTypeFloat:
            return ReadNbtFloat(reader, name, endian);
        case NbtTypeDouble:
            return ReadNbtDouble(reader, name, endian);
        case NbtTypeByteArray:
            return ReadNbtByteArray(reader, name, endian);
        case NbtTypeString:
            return ReadNbtString(reader, name, endian);
        case NbtTypeList:
            return ReadNbtList(reader, name, endian);
        case NbtTypeCompound:
            return ReadNbtCompound(reader, name, endian);
        case NbtTypeIntArray:
            return ReadNbtIntArray(reader, name, endian);
        case NbtTypeLongArray:
            return ReadNbtLongArray(reader, name, endian);
        default:
            throw "Invalid state, unknown type";
            break;
    }
}

Nbt Parse(BinaryReader& reader, Endianness endian) {
    NbtType type = ReadNbtType(reader);
    return ReadNbtSwitch(reader, endian, type, false);
}

Nbt Deserialize(uint8_t* data, uint32_t length, Endianness endian) {
    BinaryReader reader(data, length);
    return Parse(reader, endian);
}
